File: csharp/src/Ice/Connection.cs
// Copyright (c) ZeroC, Inc.

#nullable enable

namespace Ice
{
    /// <summary>
    /// The batch compression option when flushing queued batch requests.
    /// </summary>
    public enum CompressBatch
    {
        /// <summary>
        /// Compress the batch requests.
        /// </summary>
        Yes,

        /// <summary>
        /// Don't compress the batch requests.
        /// </summary>
        No,

        /// <summary>
        /// Compress the batch requests if at least one request was made on a compressed proxy.
        /// </summary>
        BasedOnProxy
    }

    public delegate void CloseCallback(Connection con);

    /// <summary>
    /// The user-level interface to a connection.
    /// </summary>
    public interface Connection
    {
        /// <summary>
        /// Aborts this connection.
        /// </summary>
        void abort();

        /// <summary>
        /// Closes the connection gracefully after waiting for all outstanding invocations to complete.
        /// </summary>
        /// <returns>A task that completes when the connection is closed.</returns>
        /// <remarks>If this operation takes longer than the configured close timeout, the connection is aborted with a
        /// <see cref="CloseTimeoutException"/>.</remarks>
        Task closeAsync();

        /// <summary>
        /// Create a special proxy that always uses this connection.
        /// This can be used for callbacks from a server to a
        /// client if the server cannot directly establish a connection to the client, for example because of firewalls. In
        /// this case, the server would create a proxy using an already established connection from the client.
        /// </summary>
        /// <param name="id">The identity for which a proxy is to be created.</param>
        /// <returns>A proxy that matches the given identity and uses this connection.</returns>
        ObjectPrx createProxy(Identity id);

        /// <summary>
        /// Associates an object adapter with this connection. When a connection receives a request, it dispatches this
        /// request using its associated object adapter. If the associated object adapter is null, the connection
        /// rejects any incoming request with an <see cref="ObjectNotExistException" />.
        /// The default object adapter of an incoming connection is the object adapter that created this connection;
        /// the default object adapter of an outgoing connection is the communicator's default object adapter.
        /// </summary>
        /// <param name="adapter">The object adapter to associate with this connection.</param>
        /// <seealso cref="Communicator.getDefaultObjectAdapter"/>
        void setAdapter(ObjectAdapter? adapter);

        /// <summary>
        /// Gets the object adapter associated with this connection.
        /// </summary>
        /// <returns>The object adapter associated with this connection.</returns>
        ObjectAdapter? getAdapter();

        /// <summary>
        /// Get the endpoint from which the connection was created.
        /// </summary>
        /// <returns>The endpoint from which the connection was created.</returns>
        Endpoint getEndpoint();

        /// <summary>
        /// Flush any pending batch requests for this connection.
        /// This means all batch requests invoked on fixed proxies associated with the connection.
        /// </summary>
        /// <param name="compress">Specifies whether or not the queued batch requests should be compressed before being sent over
        /// the wire.</param>
        void flushBatchRequests(CompressBatch compress);

        System.Threading.Tasks.Task flushBatchRequestsAsync(
            CompressBatch compress,
            System.IProgress<bool>? progress = null,
            System.Threading.CancellationToken cancel = default);

        /// <summary>
        /// Set a close callback on the connection.
        /// The callback is called by the connection when it's closed. The callback
        /// is called from the Ice thread pool associated with the connection. If the callback needs more information about
        /// the closure, it can call Connection.throwException.
        /// </summary>
        /// <param name="callback">The close callback object.</param>
        void setCloseCallback(CloseCallback callback);

        /// <summary>
        /// Return the connection type.
        /// This corresponds to the endpoint type, i.e., "tcp", "udp", etc.
        /// </summary>
        /// <returns>The type of the connection.</returns>
        string type();

        /// <summary>
        /// Returns the connection information.
        /// </summary>
        /// <returns>The connection information.</returns>
        ConnectionInfo getInfo();

        /// <summary>
        /// Set the connection buffer receive/send size.
        /// </summary>
        /// <param name="rcvSize">The connection receive buffer size.</param>
        /// <param name="sndSize">The connection send buffer size.</param>
        void setBufferSize(int rcvSize, int sndSize);

        /// <summary>
        /// Throw an exception indicating the reason for connection closure.
        /// For example,
        /// CloseConnectionException is raised if the connection was closed gracefully, whereas
        /// ConnectionAbortedException/ConnectionClosedException is raised if the connection was manually closed by
        /// the application. This operation does nothing if the connection is not yet closed.
        /// </summary>
        void throwException();
    }

    /// <summary>
    ///  Base class providing access to the connection details.
    /// </summary>
    public class ConnectionInfo
    {
        /// <summary>
        /// The underlying connection information.
        /// </summary>
        public ConnectionInfo? underlying;

        /// <summary>
        /// Whether the connection is an incoming connection (<c>true</c>) or an outgoing connection (<c>false</c>).
        /// </summary>
        public bool incoming;

        /// <summary>
        /// The name of the adapter associated with the connection.
        /// </summary>
        public string adapterName = "";

        /// <summary>
        /// The connection id.
        /// </summary>
        public string connectionId = "";
    }

    public class IPConnectionInfo : ConnectionInfo
    {
        public string localAddress = "";
        public int localPort = -1;
        public string remoteAddress = "";
        public int remotePort = -1;
    }

    public sealed class TCPConnectionInfo : IPConnectionInfo
    {
        public int rcvSize;
        public int sndSize;
    }

    public sealed class UDPConnectionInfo : IPConnectionInfo
    {
        public string mcastAddress = "";
        public int mcastPort = -1;
        public int rcvSize;
        public int sndSize;
    }

    public sealed class WSConnectionInfo : ConnectionInfo
    {
        public Dictionary<string, string> headers = [];
    }
}

File: swift/src/Ice/Connection.swift
// Copyright (c) ZeroC, Inc.

import Foundation

/// The batch compression option when flushing queued batch requests.
public enum CompressBatch: UInt8 {
    /// Yes Compress the batch requests.
    case Yes = 0
    /// No Don't compress the batch requests.
    case No = 1
    /// BasedOnProxy Compress the batch requests if at least one request was made on a compressed proxy.
    case BasedOnProxy = 2
    public init() {
        self = .Yes
    }
}

/// An `Ice.InputStream` extension to read `CompressBatch` enumerated values from the stream.
extension InputStream {
    /// Read an enumerated value.
    ///
    /// - returns: `CompressBatch` - The enumerated value.
    public func read() throws -> CompressBatch {
        let rawValue: UInt8 = try read(enumMaxValue: 2)
        guard let val = CompressBatch(rawValue: rawValue) else {
            throw MarshalException("invalid enum value")
        }
        return val
    }

    /// Read an optional enumerated value from the stream.
    ///
    /// - parameter tag: `Int32` - The numeric tag associated with the value.
    ///
    /// - returns: `CompressBatch` - The enumerated value.
    public func read(tag: Int32) throws -> CompressBatch? {
        guard try readOptional(tag: tag, expectedFormat: .Size) else {
            return nil
        }
        return try read() as CompressBatch
    }
}

/// An `Ice.OutputStream` extension to write `CompressBatch` enumerated values to the stream.
extension OutputStream {
    /// Writes an enumerated value to the stream.
    ///
    /// parameter _: `CompressBatch` - The enumerator to write.
    public func write(_ v: CompressBatch) {
        write(enum: v.rawValue, maxValue: 2)
    }

    /// Writes an optional enumerated value to the stream.
    ///
    /// parameter tag: `Int32` - The numeric tag associated with the value.
    ///
    /// parameter _: `CompressBatch` - The enumerator to write.
    public func write(tag: Int32, value: CompressBatch?) {
        guard let v = value else {
            return
        }
        write(tag: tag, val: v.rawValue, maxValue: 2)
    }
}

/// A collection of HTTP headers.
public typealias HeaderDict = [String: String]

/// Base class providing access to the connection details.
public protocol ConnectionInfo: AnyObject {
    /// The information of the underlying transport or null if there's no underlying transport.
    var underlying: ConnectionInfo? { get set }
    /// Whether or not the connection is an incoming or outgoing connection.
    var incoming: Bool { get set }
    /// The name of the adapter associated with the connection.
    var adapterName: String { get set }
    /// The connection id.
    var connectionId: String { get set }
}

/// An application can implement this interface to receive notifications when a connection closes.
///
/// This method is called by the connection when the connection is closed. If the callback needs more information
/// about the closure, it can call Connection.throwException.
///
/// - parameter _: `Connection?` The connection that closed.
public typealias CloseCallback = (Connection?) -> Void

/// The user-level interface to a connection.
public protocol Connection: AnyObject, CustomStringConvertible {

    /// Aborts this connection.
    func abort()

    /// Closes this connection gracefully after all outstanding invocations have completed.
    /// If this operation takes longer than the configured close timeout, the connection is aborted with a
    /// `CloseTimeoutException`.
    func close() async throws

    /// Create a special proxy that always uses this connection. This can be used for callbacks from a server to a
    /// client if the server cannot directly establish a connection to the client, for example because of firewalls. In
    /// this case, the server would create a proxy using an already established connection from the client.
    ///
    /// - parameter _: `Identity` The identity for which a proxy is to be created.
    ///
    /// - returns: `ObjectPrx` - A proxy that matches the given identity and uses this connection.
    func createProxy(_ id: Identity) throws -> ObjectPrx

    /// Associates an object adapter with this connection. When a connection receives a request, it dispatches this
    /// request using its associated object adapter. If the associated object adapter is null, the connection
    /// rejects any incoming request with an ObjectNotExistException.
    /// The default object adapter of an incoming connection is the object adapter that created this connection;
    /// the default object adapter of an outgoing connection is the communicator's default object adapter.
    ///
    /// - parameter _: `ObjectAdapter?` The object adapter to associate with the connection.
    func setAdapter(_ adapter: ObjectAdapter?) throws

    /// Gets the object adapter associated with this connection.
    ///
    /// - returns: `ObjectAdapter?` - The object adapter associated with this connection.
    func getAdapter() -> ObjectAdapter?

    /// Get the endpoint from which the connection was created.
    ///
    /// - returns: `Endpoint` - The endpoint from which the connection was created.
    func getEndpoint() -> Endpoint

    /// Flush any pending batch requests for this connection. This means all batch requests invoked on fixed proxies
    /// associated with the connection.
    ///
    /// - parameter _: `CompressBatch` Specifies whether or not the queued batch requests should be compressed before
    /// being sent over the wire.
    func flushBatchRequests(
        _ compress: CompressBatch
    ) async throws

    /// Set a close callback on the connection. The callback is called by the connection when it's closed. The callback
    /// is called from the Ice thread pool associated with the connection. If the callback needs more information about
    /// the closure, it can call Connection.throwException.
    ///
    /// - parameter _: `CloseCallback?` The close callback object.
    func setCloseCallback(_ callback: CloseCallback?) throws

    /// Return the connection type. This corresponds to the endpoint type, i.e., "tcp", "udp", etc.
    ///
    /// - returns: `String` - The type of the connection.
    func type() -> String

    /// Return a description of the connection as human readable text, suitable for logging or error messages.
    ///
    /// - returns: `String` - The description of the connection as human readable text.
    func toString() -> String

    /// Returns the connection information.
    ///
    /// - returns: `ConnectionInfo` - The connection information.
    func getInfo() throws -> ConnectionInfo

    /// Set the connection buffer receive/send size.
    ///
    /// - parameter rcvSize: `Int32` The connection receive buffer size.
    ///
    /// - parameter sndSize: `Int32` The connection send buffer size.
    func setBufferSize(rcvSize: Int32, sndSize: Int32) throws

    /// Throw an exception indicating the reason for connection closure. For example,
    /// CloseConnectionException is raised if the connection was closed gracefully by the peer.
    /// This operation does nothing if the connection is not yet closed.
    func throwException() throws
}

/// Provides access to the connection details of an IP connection
public protocol IPConnectionInfo: ConnectionInfo {
    /// The local address.
    var localAddress: String { get set }
    /// The local port.
    var localPort: Int32 { get set }
    /// The remote address.
    var remoteAddress: String { get set }
    /// The remote port.
    var remotePort: Int32 { get set }
}

/// Provides access to the connection details of a TCP connection
public protocol TCPConnectionInfo: IPConnectionInfo {
    /// The connection buffer receive size.
    var rcvSize: Int32 { get set }
    /// The connection buffer send size.
    var sndSize: Int32 { get set }
}

/// Provides access to the connection details of a UDP connection
public protocol UDPConnectionInfo: IPConnectionInfo {
    /// The multicast address.
    var mcastAddress: String { get set }
    /// The multicast port.
    var mcastPort: Int32 { get set }
    /// The connection buffer receive size.
    var rcvSize: Int32 { get set }
    /// The connection buffer send size.
    var sndSize: Int32 { get set }
}

/// Provides access to the connection details of a WebSocket connection
public protocol WSConnectionInfo: ConnectionInfo {
    /// The headers from the HTTP upgrade request.
    var headers: HeaderDict { get set }
}
